docs/architecture-decision-records/0020-aggregation-for-tdr-preingest.md

```diff
@@ -0,0 +1,40 @@
+# 20. Aggregation for TDR Preingest
+
+**Date:** 2024-08-14
+
+## Context
+
+The first source to ingest into DR2 was parsed court documents transferred via TDR and parsed by TRE. The upstream workflow for this type of record enforces a 1-1 relationship between a transfer consignment and a record. We're now ready to create a preingest route for standard TDR transfers.
+
+Standard TDR transfers contain multiple files within a consignment; the count and size of the files is at the discretion of the transferring user. While the Preservation System ingest process is most efficient when ingesting batches, as opposed to individual files, the size of the batches are unlikely to match the user decisions upstream of DR2; we want to design a preingest process for TDR to be most efficient for the Preservation System.
+
+## Decision
+
+We will implement an aggregation process as part of our TDR Preingest workflow, we will request TDR notify us of new records for ingest with a message per record to allow us to aggregate. Our aggregation process will use the same AWS services as our ingest process - SQS, Lambda, Step Functions, and DynamoDB.
+
+### Aggregation Design
+
+![TDR Aggregation Diagram](/docs/images/adr/0020/tdr-aggregation-diagram.png)
+
+We will extend our usage of the existing dr2-ingest-lock table (created for [`0016-generic-ingest-idempotency.md`](./0016-generic-ingest-idempotency.md)) to act as a lock table within our aggregation. We will add a `groupId` attribute and GSI on this. And a `message` attribute to store the JSON string of the message received from TDR.
+
+Messages will be received into an SQS queue with an Event Source Mapping configured to a Lambda function. The Lambda will attempt to write each message to the lock table. The Event Source Mapping will be set with a `MaximumBatchingWindowInSeconds` of 300 seconds and `BatchSize` of 10,000; the Lambda will only be invoked when the SQS poller has 10,000 messages ready, 300 seconds have elapsed since the first message was received by the poller, or the Lambda payload is 6MB. To enable processing of batches larger than 10,000 records we will cache the current `groupId` within the Lambda container and delay processing of the group for a configurable period whilst other messages can still be added. Once this period has elapsed, we will collect the messages from the lock table and use these to build a DR2 BagIt-like ingest package. The items in the lock table will be deleted as normal when the ingest process is complete.

```

Comment:
4. Added
> ...messages can still be added; the result of this will be that multiple invocations within the same Lambda container can add messages to a single aggregation group.